app/src/main/java/com/futurice/freesound/feature/search/SoundItemViewModel.java

```diff
@@ -81,12 +86,46 @@ private String getThumbnail() {
                      .map(duration -> Math.max(duration, 1));
     }
 
+    @NonNull
+    Observable<Option<Integer>> progressPercentage() {
+        return audioPlayer.getPlayerStateOnceAndStream()
+                          .switchMap(this::progressOrNothing);
+    }
+
     void openDetails() {
         navigator.openSoundDetails(sound);
     }
 
     void toggleSoundPlayback() {
-        audioPlayer.togglePlayback(sound.previews().lowQualityMp3Url());
+        audioPlayer.togglePlayback(PlaybackSource.create(Id.from(sound.id()),
+                                                         sound.previews().lowQualityMp3Url()));
+    }
+
+    @NonNull
+    private Observable<Option<Integer>> progressOrNothing(@NonNull final PlayerState playerState) {
+        return isThisSound(playerState) ?
+                getCurrentPercentage() :
+                Observable.just(Option.none());
+    }
+
+    private boolean isThisSound(@NonNull final PlayerState playerState) {
+        return playerState.source()
+                          .filter(playbackSource -> playbackSource.id().equals(Id.from(sound.id())))
+                          .isSome();
+    }
+
+    @NonNull
+    private Observable<Option<Integer>> getCurrentPercentage() {
+        return audioPlayer.getTimePositionMsOnceAndStream()
+                          .map(positionMs -> toPercentage(
+                                  positionMs,
+                                  sound.duration()))
+                          .map(Option::ofObj);
+    }
+
+    private static int toPercentage(@NonNull final Long positionMs,
```

Comment:
I've done this, but could you explain your rationale? The streams will be dealing with `Long` and `Float`, so won't changing this involve unnecessary autoboxing overhead?